delta-app/src/main/java/io/cdap/delta/app/service/AssessmentHandler.java

```diff
@@ -288,6 +285,88 @@ public void getState(HttpServiceRequest request, HttpServiceResponder responder,
     }));
   }
 
+  /*
+    Methods to manage offset and other states
+   */
+  //Looks up the delta_offset_store table for offset data.
+  @GET
+  @Path("v1/contexts/{namespace}/apps/{app}/generations/{generation}/instanceIds/{instanceid}/offset")
+  public void getOffsetData(HttpServiceRequest request, HttpServiceResponder responder,
+                           @PathParam("namespace") String namespaceName,
+                           @PathParam("app") String app,
+                           @PathParam("generation") long generation,
+                           @PathParam("instanceid") int instanceid) {
+    respond(namespaceName, responder, (namespace) -> {
+      DeltaWorkerId id = new DeltaWorkerId(new DeltaPipelineId(namespaceName, app, generation), instanceid);
+      OffsetAndSequence offset = getStateStore().readOffset(id);
+      responder.sendString(GSON.toJson(offset));
+    });
+  }
+
+  //Upserts in the delta_offset_store table for offset data.
+  @PUT
+  @Path("v1/contexts/{namespace}/apps/{app}/generations/{generation}/instanceIds/{instanceid}/offset")
+  public void saveOffsetData(HttpServiceRequest request, HttpServiceResponder responder,
+                             @PathParam("namespace") String namespaceName,
+                             @PathParam("app") String app,
+                             @PathParam("generation") long generation,
+                             @PathParam("instanceid") int instanceid) {
+    respond(namespaceName, responder, (namespace) -> {
+      DeltaWorkerId id = new DeltaWorkerId(new DeltaPipelineId(namespaceName, app, generation), instanceid);
+      String body = StandardCharsets.UTF_8.decode(request.getContent()).toString();
+      getStateStore().writeOffset(id, GSON.fromJson(body, OffsetAndSequence.class));
```

Comment:
should be decoding it into an OffsetAndSequence using a Reader, not by converting first into a String and then into the object.

Handlers should assume the user can pass in malformed input and validate all of it before proceeding. Does respond() catch exceptions decoding the request body and return a 400? Should also do some validation to make sure it's a valid offset and sequence, for example by checking that the object itself is not null and any required fields are not null.